pkg/netpol/eval/resources.go

```diff
@@ -207,52 +207,67 @@ func (pe *PolicyEngine) upsertNamespace(ns *corev1.Namespace) error {
+	if _, ok := pe.podOwnersToRepresentativePodMap[newPod.Namespace]; !ok { // add the new namespace to the map
+		pe.podOwnersToRepresentativePodMap[newPod.Namespace] = make(map[string]*k8s.Pod)
 	}
-	firstPod, ok := pe.ownersToLabelsMap[newPod.Namespace][newPod.Owner.Name]
+	firstPod, ok := pe.podOwnersToRepresentativePodMap[newPod.Namespace][newPod.Owner.Name]
 	if !ok { // add the new ownerReference with this new pod
-		pe.ownersToLabelsMap[newPod.Namespace][newPod.Owner.Name] = newPod
+		pe.podOwnersToRepresentativePodMap[newPod.Namespace][newPod.Owner.Name] = newPod
 		return nil
 	}
 	// compare the owner first pod's labels with new pod's Labels
-	return diffBetweenPodsLabels(firstPod, newPod)
+	if key, firstVal, newVal := diffBetweenPodsLabels(firstPod, newPod); key != "" {
+		return writeLabelsDifferencesError(firstPod, newPod, key, firstVal, newVal) // err
+	}
+	return nil
 }
 
-// helper: given two pods of same owner, returns an error if there are diffs between the pods' labels maps
-func diffBetweenPodsLabels(firstPod, newPod *k8s.Pod) error {
+// helper: writeLabelsDifferencesError computes the error message of the pod's different labels
+func writeLabelsDifferencesError(firstPod, newPod *k8s.Pod, key, firstVal, newVal string) error {
 	// helping vars declarations to avoid duplicates
-	errMsgPart1 := "Input Pod resources are not supported for connectivity analysis. Found Pods of the same owner but different set of labels."
-	errMsgPart2 := ""
-	keyMissingErr := " Pod %s has label %s with value %s, and Pod %s does not have label %s"
-	differentValuesErr := " Pod %s has label %s with value %s, and Pod %s has label %s with value %s"
+	ownerName := types.NamespacedName{Namespace: firstPod.Namespace, Name: firstPod.Owner.Name}.String()
 	newPodStr := types.NamespacedName{Namespace: newPod.Namespace, Name: newPod.Name}.String()
 	firstPodStr := types.NamespacedName{Namespace: firstPod.Namespace, Name: firstPod.Name}.String()
+	errMsgPart1 := "Input Pod resources are not supported for connectivity analysis. Found Pods of the same owner " +
+		ownerName + " but with different set of labels."
+	errMsgPart2 := ""
+	keyMissingErr := " Pod %s has label %s=%s, and Pod %s does not have label %s."
+	differentValuesErr := " Pod %s has label %s=%s, and Pod %s has label %s=%s."
+	switch {
+	case firstVal == "":
+		errMsgPart2 = fmt.Sprintf(keyMissingErr, newPodStr, key, newVal, firstPodStr, key)
+	case newVal == "":
+		errMsgPart2 = fmt.Sprintf(keyMissingErr, firstPodStr, key, firstVal, newPodStr, key)
+	default: // both values are not empty
+		errMsgPart2 = fmt.Sprintf(differentValuesErr, newPodStr, key, newVal, firstPodStr, key, firstVal)
+	}
+	return errors.New(errMsgPart1 + errMsgPart2)
+}
 
+// helper: given two pods of same owner, if there are diffs between the pods' labels maps returns first captured diff components,
+// i.e. the different label's key and the different values / empty val if the key is missing in one pod's labels
```

Comment:
add to function description: if there is no diff, returns empty key (with empty values)